### PR TITLE
Simplify sha3.

### PR DESCRIPTION
**Pull request by chriseth**
Split out from https://github.com/ethereum/solidity/pull/5259 and improved

### CODE REVIEW

File: libdevcore/SHA3.cpp [hunk 1]
```diff
@@ -188,55 +174,13 @@ static inline int hash(uint8_t* out, size_t outlen,
...
-/*** KECCAK FOFs ***/
-defkeccak(224)
-defkeccak(256)
-defkeccak(384)
-defkeccak(512)
-
 }
 
-bool keccak256(bytesConstRef _input, bytesRef o_output)
+h256 keccak256(bytesConstRef _input)
```
Comment (axic) [inline libdevcore/SHA3.cpp:175]: Nice! Much more cleaner.

File: libdevcore/SHA3.cpp [hunk 2]
```diff
@@ -188,55 +174,13 @@ static inline int hash(uint8_t* out, size_t outlen,
...
 {
-	// FIXME: What with unaligned memory?
-	if (o_output.size() != 32)
-		return false;
-	keccak::keccak256(o_output.data(), 32, _input.data(), _input.size());
-//	keccak::keccak(ret.data(), 32, (uint64_t const*)_input.data(), _input.size());
-	return true;
+	h256 output;
+	hash(output.data(), output.size, _input.data(), _input.size(), 200 - (256 / 4), 0x01);
```
Comment (axic) [inline libdevcore/SHA3.cpp:182]: I think it is worth leaving at least a comment explaining the parameters.

The 0x01 is the specific padding for keccak (sha3 uses 0x06) and the way the round size (or window or whatever it was) is calculated.

**Comment by axic** [comment]
Also since the file doesn't have sha3 anymore, can you as a separate commit rename it to `keccak256.c` or `keccak.c`?

File: libdevcore/SHA3.cpp (hunk 2 above)
Comment (chriseth) [reply libdevcore/SHA3.cpp:182]: hm, the params have names in the function "hash", not sure we still need to comment here.

Comment (axic) [reply libdevcore/SHA3.cpp:182]: Well the rate rate calculation is not well explained with only having constants there. At least a comment for that could make sense.

**Comment by chriseth** [comment]
Updated.

**Review by leonardoalt — APPROVED** [review]

**Comment by chriseth** [comment]
Oh wow - this was still open?

**Pull request merged**